File: src/lib/external-purchase.ts
"use client";

// ── The US-storefront external purchase link ────────────────────────────────
//
// App Review rejected 1.0.0 (3) under Guideline 3.1.1: a paying customer signs
// into the app and uses features they bought on swiftcard.me, and that content
// "isn't available to purchase using In-App Purchase". The rejection itself
// named the remedy:
//
//   "Apps on the United States storefront may link out to the default browser,
//    using buttons, external links, or other calls to action, for payment
//    mechanisms other than in-app purchase."
//
// Guideline 3.1.1(a) confirms no entitlement is needed for that on the US
// storefront. So the compliant purchase path is a button that leaves the app.
//
// TWO THINGS THIS MODULE EXISTS TO GUARANTEE
//
// 1. It really leaves the app. `swiftcard.me` is allow-listed in
//    capacitor.config.ts (the shell IS that origin), so an <a>, a target=_blank
//    or a window.open all stay inside the WKWebView, and @capacitor/browser
//    opens an in-app SFSafariViewController that merely looks like Safari.
//    Only UIApplication.open reaches the default browser, which is what Apple
//    asked for — hence the native plugin in ios/App/App/ExternalPurchase.swift.
//
// 2. It fails CLOSED. If the plugin is missing — an older shell build, or a
//    non-US storefront if availability is ever widened — this returns false and
//    the caller renders NO button. A silent fallback to an in-webview link
//    would look compliant while being exactly the violation that got us
//    rejected, and it would ship to storefronts where the link is not allowed.

import { detectNativeApp } from "@/lib/platform";

type ExternalPurchasePlugin = {
  open: (options: { url: string }) => Promise<{ opened: boolean }>;
};

function plugin(): ExternalPurchasePlugin | undefined {
  if (typeof window === "undefined") return undefined;
  return (
    window as unknown as {
      Capacitor?: { Plugins?: { ExternalPurchase?: ExternalPurchasePlugin } };
    }
  ).Capacitor?.Plugins?.ExternalPurchase;
}

/**
 * Whether a compliant external purchase link can actually be offered right now.
 *
 * False on web (the site has its own checkout) and false in any shell build
 * whose native half predates the plugin. Callers MUST gate on this rather than
 * rendering a button and hoping — see the fail-closed note above.
 */
export function canOfferExternalPurchase(): boolean {
  return detectNativeApp() && !!plugin();
}

/** Where the link goes. `src` is for acquisition attribution, nothing more. */
export function externalPurchaseUrl(path = "/upgrade"): string {
  return `https://swiftcard.me${path}${path.includes("?") ? "&" : "?"}src=ios_link`;
}

/**
 * Open the subscribe page in the default browser.
 *
 * Resolves false when the plugin is absent or the system declined, so a caller
 * can surface a real failure instead of leaving the user staring at a button
 * that did nothing.
 */
export async function openExternalPurchase(path?: string): Promise<boolean> {
  const p = plugin();
  if (!p) return false;
  try {
    const { opened } = await p.open({ url: externalPurchaseUrl(path) });
    return !!opened;
  } catch {
    return false;
  }
}

File: ios/App/App/ExternalPurchase.swift
import Foundation
import Capacitor
import UIKit

// ─────────────────────────────────────────────────────────────────────────────
// ExternalPurchase — opens the subscribe page in the DEFAULT BROWSER.
//
// WHY A NATIVE PLUGIN AND NOT A LINK
//
// capacitor.config.ts allow-lists `swiftcard.me` under server.allowNavigation,
// because the whole shell IS that origin. Every ordinary route through the
// purchase page therefore stays inside the WKWebView:
//
//   • <a href="https://swiftcard.me/upgrade">        → in-webview navigation
//   • <a target="_blank">                            → allow-listed host, same
//   • window.open(url)                               → same
//   • @capacitor/browser Browser.open()              → SFSafariViewController,
//                                                      which is an IN-APP sheet
//
// None of those leave the app, and the last one is the trap: an SFSafari sheet
// looks like Safari and is not. App Review's rejection wording is specific —
// apps on the United States storefront "may link out to the DEFAULT BROWSER".
// UIApplication.open is the only call that does that, so the compliant path has
// to cross the bridge.
//
// WHY THIS IS ALLOWED AT ALL
//
// Guideline 3.1.1(a): "These entitlements are not required for developers to
// include buttons, external links, or other calls to action in their United
// States storefront apps." The StoreKit External Purchase Link Entitlement is a
// non-US mechanism; on the US storefront the link needs no entitlement and no
// modal sheet. This is why the app must ship US-only while this is the purchase
// path — in any other storefront the same button is a 3.1.1 violation, not a
// remedy. If availability is ever widened, this plugin must be gated on the
// StoreKit storefront country or replaced by real In-App Purchase.
//
// The web half is src/lib/external-purchase.ts, which fails CLOSED: if this
// plugin is missing (an older build, or the widened-availability case above) it
// renders no button at all rather than falling back to an in-webview link that
// would look compliant and not be.
// ─────────────────────────────────────────────────────────────────────────────

@objc(ExternalPurchasePlugin)
public class ExternalPurchasePlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "ExternalPurchasePlugin"
    public let jsName = "ExternalPurchase"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "open", returnType: CAPPluginReturnPromise)
    ]

    // Only our own origin — plus Apple's subscription-management page — may be
    // opened. The URL arrives from web code running in a remote-origin webview,
    // so it is not inherently trusted input: without this an injected string
    // could turn a compliance affordance into an arbitrary-URL opener.
    //
    // apps.apple.com is here for manageIapSubscription() (lib/iap.ts): Apple-
    // billed subs are canceled on the App Store's own page, and iOS hands that
    // URL to the system subscription sheet. Before it was allow-listed, the
    // "Manage subscription" button resolved a rejection into a swallowed catch
    // — a silently dead button, in front of exactly the reviewer who just
    // sandbox-purchased.
    private static let allowedHosts: Set<String> = ["swiftcard.me", "www.swiftcard.me", "apps.apple.com"]

    @objc func open(_ call: CAPPluginCall) {
        guard let raw = call.getString("url"),
              let url = URL(string: raw),
              let scheme = url.scheme?.lowercased(),
              scheme == "https",
              let host = url.host?.lowercased(),
              Self.allowedHosts.contains(host)
        else {
            call.reject("A https swiftcard.me url is required")
            return
        }

        DispatchQueue.main.async {
            // `options: [:]` with a completion handler, not the deprecated
            // openURL(_:) — the deprecated form is a no-op from an extension
            // context and gives no success signal, so a failure to open would
            // resolve as if the browser had launched.
            UIApplication.shared.open(url, options: [:]) { opened in
                if opened {
                    call.resolve(["opened": true])
                } else {
                    call.reject("The system declined to open the url")
                }
            }
        }
    }
}
